src/cast_core/src/AssertiveActor.cpp

```diff
@@ -0,0 +1,216 @@
+    } else if (type != actualVal.type()) {
+        BOOST_LOG_TRIVIAL(info) << prefix(config) << "Types differ";
+        return false;
+    } else if (type == bsoncxx::type::k_array) {
+        bsoncxx::array::view expectedArr{expectedVal.get_array().value};
+        bsoncxx::array::view actualArr{actualVal.get_array().value};
+        return equalBSONArrays(config, expectedArr, actualArr);
+    } else if (type == bsoncxx::type::k_document) {
+        bsoncxx::document::view expectedDoc{expectedVal.get_document().value};
+        bsoncxx::document::view actualDoc{actualVal.get_document().value};
+        return equalBSONDocs(config, expectedDoc, actualDoc);
+    }
+    return actualVal.get_value() == expectedVal.get_value();
+}
+
+bool equalBSONDocs(AssertiveActor::PhaseConfig* config, const bsoncxx::document::view& expected, const bsoncxx::document::view& actual) {
+    for (auto actualVal : actual) {
+        auto key = actualVal.key();
+        if (key == "_id") {
+            continue;
+        }
+        if (key == "num") {
+            continue;
+        }
+        auto expectedVal = expected[key];
+        if (!equal<bsoncxx::document::element>(config, expectedVal, actualVal)) {
+            BOOST_LOG_TRIVIAL(info) << prefix(config) << "Documents differ in field: " << key;
+            return false;
+        }
+    }
+    for (auto expectedVal: expected) {
+        auto key = expectedVal.key();
+        if (key == "_id") {
+            continue;
+        }
+        if (key == "num") {
+            continue;
+        }
+        if (!actual[key]) {
+            BOOST_LOG_TRIVIAL(info) << prefix(config) << "Documents differ in field: " << key;
+            return false;
+        }
+    }
+    return true;
+}
+
+bool equalBSONArrays(AssertiveActor::PhaseConfig* config, const bsoncxx::array::view& expected, const bsoncxx::array::view& actual) {
+    auto expectedArrLen = std::distance(expected.begin(), expected.end());
+    auto actualArrLen = std::distance(actual.begin(), actual.end());
+    if (expectedArrLen != actualArrLen) {
+        BOOST_LOG_TRIVIAL(info) << prefix(config) << "Arrays differ in length. Expected " << expectedArrLen << " but found " << actualArrLen;
+        return false;
+    }
+
+    auto expectedIt = expected.begin();
+    int count = 0;
+    for (auto actualVal : actual) {
+        auto expectedVal = *expectedIt;
+        if (!equal<bsoncxx::array::element>(config, expectedVal, actualVal)) {
+            BOOST_LOG_TRIVIAL(info) << prefix(config) << "Arrays differ in idx " << count;
+            return false;
+        }
+        ++expectedIt;
+        ++count;
+    }
+
+    return true;
+}
+
+void AssertiveActor::run() {
+    for (auto&& config : _loop) {
+        for (const auto&& _ : config) {
+            auto expected = config->expectedExpr();
+            auto actual = config->actualExpr();
+
+            auto assertOp = _assert.start();
+            try {
```

Comment:
Should we add some more granular nested metrics collection here in addition to the total? Maybe a metric for getting the expected and a metric for getting the actual? Would it make sense to time the comparison too? I could imagine for large datasets that could be a nontrivial perf test itself.

We do similar for e.g. TotalBulkInsert, where we time the total insert and also each individual one.

Totally up to you, whether there might be meaningful value for it.

Reply:
Sounds good!